drivers: axi: Add platform agnostic spi engine

scuciurean opened the pull request:
Added platform-agnostic spi engine inside xilinx platform drivers.

cristiantpop commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+ * @param fifo Command buffer, ussualy used in fifo mode
+ * @param cmd Command to be added
+ */
+void spi_engine_queue_append_cmd(struct spi_engine_cmd_queue **fifo,
```
Add static to all local functions.

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+/**
+ * @brief Get the command from the end of the queue
+ *
+ * @param fifo Command buffer, ussualy used in fifo mode
```
usually

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	int32_t ret;
+
+	/* Check if offload is enabled */
+	if(desc->offload_config & (OFFLOAD_TX_EN | OFFLOAD_RX_EN)) {
```
`desc` should be tested before dereferencing. This applies to all subsequent functions that use desc as parameter.

mchindri commented on drivers/axi_core/spi_engine/spi_engine.h:
```diff
@@ -0,0 +1,188 @@
+int32_t spi_engine_remove(spi_desc *desc);
+
+/* Initialize the SPI engine offload module */
+int32_t spi_engine_offload_init(struct spi_desc *desc,
```
Could this be inside spi_engine_init and pass `spi_engine_offload_init_param` as a field of spi_engine_init_param?
In the case the field is null make no offload init.
I'am asking this because there is no spi_engine_offload_remove and the remove for this is done in spi_engine_remove.
Also it seems the spi_engine_offload_transfer uses the spi_desc not an descriptor from offload_init. It looks a little puzzling to me.

scuciurean replied:
There are two structures (spi_engine and spi_engine_offload) because there are two independent modes in which the engine can work.

mchindri commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+ * @param bytes_number The number of bytes to be cnoverted
+ * @return uint8_t A number of words in which bytes_number can be grouped
+ */
+uint8_t spi_get_words_number(struct spi_engine_desc *desc,
```
could this be an static inline function?

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+
+	desc_extra = desc->extra;
+
+	engine_command = (cmd >> 12) & 0x0F;
```
As somebody who is not familiar with the code, this looks obscure. Please document why this shifting and masking is done.

scuciurean replied:
The command structure is documented in the spi engine's wiki. I'll add a link in the coment above this command.

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	desc_extra = desc->extra;
+
+	/* Set the data transfer length */
+	spi_engine_queue_append_cmd(&msg->cmds,
```
`msg` needs to be tested before dereferencing

mchindri replied:
It is an internal function. I think here msg must always come well configure.

buha replied:
Well, in this case the function should be documeted accordingly, something like "this function assumes non-NULL input msg parameter". At some point during development, it is possible that an internal function becomes external, and guess what, the developer won't check the contents and how the pointers are not dereferenced, it will simply remove 'static' qualifier, and you end up with a bug.

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	do {
+		spi_engine_queue_lifo_get_cmd(&msg->cmds, &data);
+		spi_engine_write_cmd(desc, data);
+		usleep(1000);
```
This looks like a hack to make it work. Is this really needed ? If yes, please document why is it needed.

scuciurean replied:
This was forgotten after some tests and will be removed.

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	if (!desc || !eng_desc)
+		return FAILURE;
+
+	spi_engine_init = param->extra;
```
`param` needs to be tested before dereferencing

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+
+	/* Pack the bytes into engine WORDS */
+	for (i = 0; i < bytes_number; i++)
+		msg->tx_buf[i / word_len] |= data[i] << (desc_extra->data_width-
```
`desc_extra` needs to be tested before dereferencing

buha commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	msg->length = words_number;
+
+	/* Init the rx and tx buffers with 0s */
+	for (i = 0; i < words_number; i++) {
```
you could use bzero() for this

mchindri commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	uint8_t 		word_len;
+	uint8_t 		words_number;
+	int32_t 		ret;
+	struct spi_engine_msg	*msg;
```
this haven't to be a pointer. This way there will not be need for  malloc and free

mchindri replied:
same for spi_engine_offload_transfer

scuciurean replied:
I would rather pass them as pointers and allocate the memory inside the spi engine driver instead of constraining the user to allocate memory in his application.

mchindri commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	words_number = spi_get_words_number(desc_extra, bytes_number);
+
+	msg = (spi_engine_msg *)malloc(sizeof(*msg));
+	msg->cmds = (spi_engine_cmd_queue*)malloc(sizeof(*msg->cmds));
```
instead of this will be better something like 
spi_engine_queue_init(&msg->cmds);

mchindri replied:
same for spi_engine_offload_transfer

scuciurean replied:
This was discussed in the previous comment.

mchindri commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	if (!msg->cmds || !msg)
+		return FAILURE;
+
+	msg->tx_buf =(uint32_t*)malloc(words_number * sizeof(msg->tx_buf[0]));
```
you can use calloc and it will fill the buffers with 0.
Also calloc it is prefered in all the allocations.

scuciurean commented:
Changelog -> V2:

1) Add static specifier to all local functions.

mchindri commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+
+	if(offload_en) {
+		for(i = 0; i < desc_extra->offload_tx_len; i++) {
+			spi_engine_write(desc_extra,
```
there is no need for {}. The same for others calls to spi_engine_write

mchindri commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+
+	word_length = spi_get_word_lenght(eng_desc);
+	if(eng_desc->offload_config & OFFLOAD_TX_EN) {
+		axi_dmac_transfer(eng_desc->offload_tx_dma,
```
no need for {}. Same for next if

mchindri commented on drivers/platform/xilinx/spi.c:
```diff
@@ -70,116 +157,79 @@ int32_t spi_init(struct spi_desc **desc,
-	xdesc = (struct xil_spi_desc *)malloc(sizeof(*xdesc));
-	if(!sdesc || !xdesc) {
+	sdesc = malloc(sizeof(*sdesc));
+	if(!sdesc) {
```
you must return only FAILURE, no free needed

mchindri commented on drivers/platform/xilinx/spi.c:
```diff
@@ -70,116 +157,79 @@ int32_t spi_init(struct spi_desc **desc,
+	if (*spi_type == SPI_ENGINE) {
+		edesc = malloc(sizeof(spi_engine_desc));
+		if(!edesc) {
+			free(edesc);
```
free(sdesc). And the next one is not ok either. Check the goto failure pattern here https://github.com/analogdevicesinc/no-OS/blob/master/drivers/platform/aducm3029/spi.c#L105

cristiantpop commented on drivers/platform/xilinx/spi.c:
```diff
@@ -54,11 +54,98 @@
+	return FAILURE;
+}
+
+int32_t spi_init_ps(struct xil_spi_desc *xdesc,
```
Missing function comment.

cristiantpop commented on drivers/platform/xilinx/spi.c:
```diff
@@ -54,11 +54,98 @@
 /************************ Functions Definitions *******************************/
 /******************************************************************************/
 
+int32_t spi_init_pl(struct xil_spi_desc *xdesc,
```
Missing function comment.

mchindri commented on drivers/axi_core/spi_engine/spi_engine.h:
```diff
@@ -0,0 +1,188 @@
+/******************************************************************************/
+
+/* Initialize the SPI engine device */
+int32_t spi_engine_init(struct spi_desc **desc,
```
I think a struct spi_engine_desc must be initialized which will be the extra field of the spi_desc. Review the init and remove flow from spi.c and you will see the problem.

scuciurean replied:
In this function, I need to access members from the spi_init_param so, I need a spi_desc to be passed.

mchindri commented on drivers/platform/xilinx/spi.c:
```diff
@@ -54,11 +54,98 @@
+ps_error:
+	free(xdesc->instance);
+#endif
+	free(xdesc);
```
this free must not be done here. It is done in the outside function. Same for init_pl

scuciurean replied:
I believe you are suggesting that the user should check the init and call the remove function himself in case of error. In this way, again, we are forcing the user to call additional functions.

mchindri commented:
f53eaba18faf9bda0648ebfaa4d5345d42355592 this comment should be splited in 2. First create the functions spi_init_ps and spi_init_pl and the support for spi_engine.

adrimbarean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+ * 		- 24
+ * 		- 32
+ */
+static int32_t spi_engine_set_transfer_wdith(struct spi_desc *desc,
```
nitpick: wdith -> width

adrimbarean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+
+	if (data_wdith > desc_extra->max_data_width) {
+		desc_extra->data_width = desc_extra->max_data_width;
+		dev_dbg(desc, "Data width (%d) exceeds the maximum of %d bits",
```
Maybe the message here should specifically state that the data width has been changed  to the maximum possible. The user might assume nothing changed based on the error message.

adrimbarean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+ * @param fifo Command buffer, ussualy used in fifo mode
+ * @param cmd Command to be added
+ */
+static void spi_engine_queue_add_cmd(struct spi_engine_cmd_queue **fifo,
```
I don't think double pointer is necessary here. `fifo` should not change.

scuciurean replied:
If I would use spi_engine_queue_append_cmd, the new element would be added before the first one in the queue (fifo) so the address of the resulting queue would change to the newly added element.

adrimbarean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+		local_fifo = local_fifo->next;
+	}
+	*cmd = local_fifo->cmd;
+	if(last_fifo->next != NULL) {
```
Is this deferencing a NULL pointer if `fifo` has only one element?

adrimbarean replied:
Also it seems like `last_fifo` will always be the second to last element in the FIFO, so `last_fifo->next` would never equal NULL.

scuciurean replied:
After reading your comment I realized that I'm using a weird last->next state instead of just using the current address.

adrimbarean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+ * @param cmd Command with wich the buffer will be initiated
+ * @return int32_t Failure if the memory allocation failed
+ */
+static int32_t spi_engine_queue_new_cmd(struct spi_engine_cmd_queue **fifo,
```
nitpick: the command queue here behaves more like a directed graph than a FIFO, so probably another parameter name would be appropriate.

scuciurean replied:
This queue has the possibility to add and remove items in both fifo and lifo modes. The name was chosen because in most cases is used as a fifo.

adrimbarean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,960 @@
+
+
+	if ((*fifo)->next) {
+		local_fifo = (*fifo);
```
nitpick: No need for parenthesis.

scuciurean commented on drivers/platform/xilinx/spi.c:
```diff
@@ -192,26 +242,35 @@ int32_t spi_remove(struct spi_desc *desc)
 	case SPI_PS:
 #ifdef XSPIPS_H
-
+		xdesc = desc->extra;
```
It's indented by the case above

scuciurean commented on drivers/platform/xilinx/spi.c:
```diff
@@ -192,26 +242,35 @@ int32_t spi_remove(struct spi_desc *desc)
 	case SPI_ENGINE:
 #ifdef SPI_ENGINE_H
-
+		spi_engine_remove(desc);
```
It's indented by the case above

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+{
+	dev_dbg_reg(desc, "reg_addr = 0x%X; reg_data = 0x%X", reg_addr,
+		    reg_data);
+	axi_io_write(desc->spi_engine_baseaddr, reg_addr, reg_data);
```
Since the descriptor is tested in the init function , there is no need to verify it each time

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+
+	local_fifo->cmd = cmd;
+	local_fifo->next = NULL;
+	*fifo = local_fifo;
```
I do this in the spi_engine_write_and_read once and the other cases in which fifo may be null is when a new queue is allocated. In that case, in spi_engine_queue_new_cmd the new fifo is verified

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+ * @param fifo Command buffer, ussualy used in fifo mode
+ * @param cmd Command to be added
+ */
+void spi_engine_queue_add_cmd(struct spi_engine_cmd_queue **fifo,
```
spi_engine_queue is used only in spi_engine.c and there is no need for a separate file

scuciurean commented on include/util.h:
```diff
@@ -85,12 +85,7 @@
-		t = t >> (BITS_PER_LONG - (h + 1));						\
-		t;														\
-})
+#define GENMASK(h, l)	(((~0UL) << (l)) & (~0UL >> (32 - 1 - (h))))
```
It's the same as include/linux/bits.h

scuciurean commented on drivers/axi_core/spi_engine/spi_engine_private.h:
```diff
@@ -0,0 +1,201 @@
+/******************************************************************************/
+
+#ifdef DEBUG_LEVEL
+#define dev_dbg_reg(dev, format, ...){			\
```
Good point

scuciurean commented on drivers/axi_core/spi_engine/spi_engine_private.h:
```diff
@@ -0,0 +1,201 @@
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+typedef struct spi_engine_cmd_queue {
```
spi_engine_queue is used only in spi_engine.c and there is no need for a separate file

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#ifndef _SPI_ENGINE_
```
This was copied by mistake among with the copyrights

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+
+/**
+ * @brief Static variable used to sync transfers
+ *
```
This was generated by a doxygen plugin that i use. Maybe it's a different format.

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	eng_desc = (struct spi_engine_desc*)malloc(sizeof(*eng_desc));
+
+	if (!desc || !eng_desc)
+		return FAILURE;
```
In my opinion, the check after each allocation seems to be redundant since in the application you free (remove) the descriptor in case of init failure.

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	local_desc = (struct spi_desc *)malloc(sizeof(*local_desc));
+	eng_desc = (struct spi_engine_desc*)malloc(sizeof(*eng_desc));
+
+	if (!desc || !eng_desc)
```
Right

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+	/* Get current data width */
+	spi_engine_read(eng_desc, SPI_ENGINE_REG_DATA_WIDTH, &data_width);
+	eng_desc->max_data_width = data_width;
+	dev_dbg(desc, "Data with is 8 bits (max %d)", eng_desc->max_data_width);
```
Yes indeed, the 8 was left behind from a previous implementation.

scuciurean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,958 @@
+				const struct spi_engine_offload_init_param *param)
+{
+	struct spi_engine_desc	*eng_desc;
+	struct axi_dmac_init	*dmac_init;
```
I would rather pass them as pointers and allocate the memory inside the spi engine driver instead of constraining the user to allocate memory in his application.

scuciurean replied:
Sorry, I've missed that init

scuciurean commented:
Changelog -> V3:

1. Added function comments in spi.c
2. Removed the delay between the command writes
3. Now buffer zeroing is done with calloc.
4. Enclosed debug messages inside a while loop
5. Check pointers before dereferencing
6. Removed unnecessary memory allocations.

scuciurean commented:
Changelog -> V4:

1. Applied astyle

cristiantpop commented on drivers/platform/xilinx/spi.c:
```diff
@@ -54,11 +54,112 @@
+ * @param xinit Platform specific SPI init param
+ * @return int32_t FAILURE if the device couldn't be found or configured
+ */
+int32_t spi_init_ps(struct xil_spi_desc *xdesc,
```
add static

scuciurean commented:
Changelog -> V5:

1. Add more function comments.
2. Add static specifier to internal functions.

scuciurean commented:
Changelog -> V6:

1. Removed mask macro from the read/write instructions in order to make possible the initialization of the user commands inside an array (arrays can't be initialized with values containing arithmetic operators)

scuciurean commented:
Changelog -> V6:

1. Remove debug macros
2. Assign dev extra to the descriptor in xil_spi.c

scuciurean commented:
Changelog -> V6:

scuciurean commented:
Changelog -> V8:

1. Add pointer differencing in spi_remove.

scuciurean commented:
Changelog -> V9:

1. Cleanup: removed some intermediary pointers.

scuciurean commented:
Changelog -> V10:

1. Add include check in spi.c in case spi_engine is not used.

scuciurean commented:
Changelog -> V11:

1. Removed extra paranthesis in cloc_div computing.

mchindri commented on drivers/platform/xilinx/spi.c:
```diff
@@ -54,166 +54,235 @@
 #ifdef SPI_ENGINE_H
-		//TODO: Implement SPI engine feature
-
+		ret = spi_engine_init(desc, param);
```
makes more sense if the header will be 
`int32_t spi_engine_init(struct spi_desc *desc, const struct spi_init_param *param)`
because the general descriptor is not allocated inside of these inits.
And the call to this functions will be:
```
ret = spi_init_pl(*desc, param);
ret = spi_init_ps(*desc, param);
ret = spi_engine_init(*desc, param);
```

mchindri commented on drivers/axi_core/spi_engine/spi_engine.h:
```diff
@@ -0,0 +1,192 @@
+	uint32_t		cs_delay;
+	/** Data with of one SPI transfer ( in bits ) */
+	uint8_t			data_width;
+} spi_engine_init_param;
```
Just as reminder: to remove `spi_engine_init_param` before } . Same for next

mchindri commented on drivers/axi_core/spi_engine/spi_engine.h:
```diff
@@ -0,0 +1,192 @@
+				  uint8_t bytes_number);
+
+/* Free the resources used by the SPI engine device */
+int32_t spi_engine_remove(spi_desc *desc);
```
Just as reminder: use `struct spi_desc`

mchindri replied:
you forgot this.

adrimbarean commented on drivers/axi_core/spi_engine/spi_engine.c:
```diff
@@ -0,0 +1,838 @@
+
+	spi_engine_init = param->extra;
+
+	(*desc)->max_speed_hz = param->max_speed_hz;
```
I think *desc needs to be checked here. Usually memory allocation is done in the init and I assume that spi_engine_init should be called within or after the spi_ini method. But since it's an API, I think we should guard against it.

scuciurean commented:
Changelog -> V12:

1. Remove struct initializers.
2. Check descriptor pointer inside spi_engine_init

mchindri commented on drivers/platform/xilinx/spi.c:
```diff
@@ -54,166 +54,235 @@
-	return SUCCESS;
+	if (ret != SUCCESS) {
+init_error:
+		free(desc);
```
desc is of type struct spi_desc **
you must free(*desc)

scuciurean replied:
I missed that two times somehow.

scuciurean commented:
Changelog -> V13:

1. Correct memory freeing.

mchindri commented:
https://github.com/analogdevicesinc/no-OS/pull/437#discussion_r378774660
You forgot this
